chain/store/name.go

```diff
@@ -1,63 +1,178 @@
 package store
 
 import (
-	"strings"
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/nknorg/nkn/chain/trie"
+
+	"github.com/nknorg/nkn/common/serialization"
 )
 
-func getRegistrantId(registrant []byte) string {
-	return string(registrant)
+type nameInfo struct {
+	registrant []byte
+	expiresAt  uint32
 }
 
-func getNameId(name string) string {
-	return strings.ToLower(name)
+type namesCleanup map[string]struct{}
+
+func newNameInfo(registrant []byte, expiresAt uint32) *nameInfo {
+	return &nameInfo{registrant: registrant, expiresAt: expiresAt}
 }
 
-func (sdb *StateDB) updateName(registrant []byte, name string) error {
-	registrantId := getRegistrantId(registrant)
-	nameId := getNameId(name)
+func (ni *nameInfo) Serialize(w io.Writer) error {
+	if err := serialization.WriteVarBytes(w, ni.registrant); err != nil {
+		return err
+	}
+	if err := serialization.WriteUint32(w, ni.expiresAt); err != nil {
+		return err
+	}
+	return nil
+}
 
-	err := sdb.trie.TryUpdate(append(NamePrefix, nameId...), registrant)
+func (ni *nameInfo) Deserialize(r io.Reader) error {
+	var err error
+	ni.registrant, err = serialization.ReadVarBytes(r)
 	if err != nil {
 		return err
 	}
+	ni.expiresAt, err = serialization.ReadUint32(r)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-	return sdb.trie.TryUpdate(append(NameRegistrantPrefix, registrantId...), []byte(name))
+func (ni *nameInfo) Bytes() ([]byte, error) {
+	buff := bytes.NewBuffer(nil)
+	err := ni.Serialize(buff)
+	if err != nil {
+		return nil, err
+	}
+	return buff.Bytes(), nil
 }
 
-func (sdb *StateDB) setName(registrant []byte, name string) {
-	registrantId := getRegistrantId(registrant)
-	nameId := getNameId(name)
+func (ni *nameInfo) String() string {
```

Comment:
See comments below, `String()` method is no longer needed if we remove `getRegName` method

Reply:
@yilunzhang We need two API of an object struct:
One is machine-oriented `obj.Bytes()` for deserialization and `consistent hash`
Another is human-oriented `obj.String()` for logger, debug, rpc response etc.

Reply:
I think it's better to add them together with other similar structure like pubsub, otherwise only nameInfo has this method